Library/Homebrew/cask/lib/hbc/cask.rb

```diff
@@ -10,6 +10,13 @@ class Cask
     def initialize(token, sourcefile_path: nil, &block)
       @token = token
       @sourcefile_path = sourcefile_path
+
+      @tap = if @sourcefile_path == CaskLoader.default_path(@token)
+        Hbc.default_tap
+      elsif @sourcefile_path.to_s =~ HOMEBREW_TAP_PATH_REGEX
+        Tap.fetch(Regexp.last_match(1), Regexp.last_match(2))
+      end
```

Comment:
I don't think this will ever work this way with `brew cask list`. For installed Casks, the `sourcefile_path` will always be the one in the `Caskroom` metadata directory, hence will never contain any information about the Tap.